### ISSUE TITLE
Bug: Unable to bypass TS2408 error: Setters cannot return a value

### ISSUE THREAD

**Issue by eligrey** [microsoft/TypeScript#61561]
### 🔎 Search Terms

This is a dupe of <https://github.com/microsoft/TypeScript/issues/13216>, which was incorrectly closed with the invalid reason "We can't change the fact that setters don't return a value. This is just how JavaScript works"

JavaScript accessor setters can indeed return values, and people need to be able to write code that returns values inside accessors for compatibility reasons. I think that this is a bad code practice that linters should flag by default, but this shouldn't emit a TypeScript error that requires syntax changes to mitigate.

### 🕗 Version & Regression Information

This is the behavior in every version I tried, and I reviewed the FAQ for relevant entries

### 💻 Code

```ts
const foo = {
  set bar(_) {
    return 'return value';
  }
};
console.log(`setter can${Object.getOwnPropertyDescriptor(foo, 'bar').set() !== 'return value' ? "'t" : ''} return values`)
```

### 🙁 Actual behavior

TypeScript compiler emits an error

### 🙂 Expected behavior

TypeScript compiler does not emit an error

### Additional information about the issue

This issue can be bypassed by replacing instances of setter shorthand syntax with calls to `Object.defineProperty` or `Object.defineProperties`

**Comment by RyanCavanaugh** [comment]
All "errors" are called "errors" because we don't have anything called a "warning", since the definition of what's an error and what's a warning is subject to endless debate. You're always free to ignore an "error" with `@ts-ignore`.

**Comment by eligrey** [comment]
@RyanCavanaugh Ah, thank you for the clarification! I was able to bypass the error as such:

```ts
const foo = {
  set bar(_) {
    /* eslint-disable no-setter-return, @typescript-eslint/ban-ts-comment */
    // @ts-ignore
    return 'return value';
    /* eslint-enable no-setter-return, @typescript-eslint/ban-ts-comment */
  }
};
const settersCanReturnValues = Object.getOwnPropertyDescriptor(foo, 'bar').set() === 'return value';
console.log(`setters can${settersCanReturnValues? '' : "'t"} return values`);
```

Should I close this issue, or do you think it's worth re-scoping to improve the DX here?

**Comment by eligrey** [comment]
@RyanCavanaugh notably, set accessors cannot have return types, which is a related bug that causes actual compilation issues.

e.g.

```ts
const foo = {
  set bar(_): string {
    /* eslint-disable no-setter-return, @typescript-eslint/ban-ts-comment */
    // @ts-ignore
    return 'return value';
    /* eslint-enable no-setter-return, @typescript-eslint/ban-ts-comment */
  }
};
const settersCanReturnValues = Object.getOwnPropertyDescriptor(foo, 'bar').set() === 'return value';
console.log(`setters can${settersCanReturnValues? '' : "'t"} return values`);
```

**Comment by RyanCavanaugh** [comment]
> Should I close this issue, or do you think it's worth re-scoping to improve the DX here?

I would take a *very* tactical change to the error message if someone comes up with a message that's equivalently good at conveying the point in the common case (which I think is just a person thinking they're in the getter when they're in the setter, or thinking they're inside a closure but not), but it's probably just best left alone.

**Comment by eligrey** [comment]
I re-scoped this to the return types issue; feel free to close or further re-scope as a feature request
